lang/java/reef-io/src/main/java/org/apache/reef/io/data/loading/impl/AbstractEvaluatorToPartitionStrategy.java

```diff
@@ -0,0 +1,256 @@
+    for (final String serializedDataPartition : serializedDataPartitions) {
+      final DistributedDataSetPartition dp = DistributedDataSetPartitionSerializer.deserialize(serializedDataPartition);
+      final ExternalConstructor<JobConf> jobConfExternalConstructor = new JobConfExternalConstructor(
+          inputFormatClassName, dp.getPath());
+      try {
+        final JobConf jobConf = jobConfExternalConstructor.newInstance();
+        final InputFormat inputFormat = jobConf.getInputFormat();
+        final InputSplit[] inputSplits = inputFormat.getSplits(jobConf, dp.getDesiredSplits());
+        if (LOG.isLoggable(Level.FINEST)) {
+          LOG.log(Level.FINEST, "Splits for partition: {0} {1}", new Object[] {dp, Arrays.toString(inputSplits)});
+        }
+        this.totalNumberOfSplits += inputSplits.length;
+        splitsPerPartition.put(dp, inputSplits);
+      } catch (final IOException e) {
+        throw new RuntimeException("Unable to get InputSplits using the specified InputFormat", e);
+      }
+    }
+    init(splitsPerPartition);
+    LOG.log(Level.FINE, "Total Number of splits: {0}", this.totalNumberOfSplits);
+  }
+
+  /**
+   * Initializes the locations of the splits where we'd like to be loaded into.
+   * Sets all the splits to unallocated
+   *
+   * @param splitsPerPartition
+   *          a map containing the input splits per data partition
+   */
+  private void init(final Map<DistributedDataSetPartition, InputSplit[]> splitsPerPartition) {
+    final Pair<InputSplit[], DistributedDataSetPartition[]>
+                                      splitsAndPartitions = getSplitsAndPartitions(splitsPerPartition);
+    final InputSplit[] splits = splitsAndPartitions.getFirst();
+    final DistributedDataSetPartition[] partitions = splitsAndPartitions.getSecond();
+    Validate.isTrue(splits.length == partitions.length);
+    for (int splitNum = 0; splitNum < splits.length; splitNum++) {
+      LOG.log(Level.FINE, "Processing split: " + splitNum);
+      final InputSplit split = splits[splitNum];
+      final NumberedSplit<InputSplit> numberedSplit = new NumberedSplit<InputSplit>(split, splitNum,
+          partitions[splitNum]);
+      unallocatedSplits.add(numberedSplit);
+      updateLocations(split, numberedSplit);
+    }
+    if (LOG.isLoggable(Level.FINE)) {
+      for (final Map.Entry<String, BlockingQueue<NumberedSplit<InputSplit>>> locSplit : locationToSplits.entrySet()) {
+        LOG.log(Level.FINE, locSplit.getKey() + ": " + locSplit.getValue().toString());
+      }
+    }
+  }
+
+  /**
+   * Each strategy should update the locations where they want the split to be
+   * loaded into. For example, the split physical location, certain node,
+   * certain rack
+   *
+   * @param split
+   *          the inputSplit
+   * @param numberedSplit
+   *          the numberedSplit
+   */
+  protected abstract void updateLocations(final InputSplit split, final NumberedSplit<InputSplit> numberedSplit);
```

Comment:
The `split` is also returned by `numberedSplit.getEntry()`, right? If so, why do we need both parameters?